Start sessions automatically after device boot.

hugorosario opened the pull request:
Implementation of requested features in issue https://github.com/CypherpunkArmory/UserLAnd/issues/368

This changes introduce the ability to start sessions automatically after Android boots.

A BootReceiver was created to handle the ACTION_BOOT_COMPLETED broadcast which the  system sends after boot.
Upon receiving the intent, it starts the ServerService with the "autostart" action which will instruct the service to start every session that has the startOnBoot flag and execute the "/support/autostart.sh" script which the user can create to add commands.

It also introduces CPU and WIFI wakelocks which can be enabled/disabled from the foreground notification. (very similar to what Termux does).

MatthewTighe commented:
Thanks for your contribution! I'll work on reviewing this today.

MatthewTighe commented on app/src/main/java/tech/ula/ServerService.kt:
```diff
@@ -80,17 +85,100 @@ class ServerService : Service() {
+     * Used in conjunction with manifest attribute `android:stopWithTask="true"`
+     * to clean up when app is swiped away.
+     *
+     * Warning : stopWithTask on the manifest is now false because if true, this method is not called
```
You can remove the unnecessary parts of this comment.

hugorosario replied:
Will do

hugorosario commented on app/src/main/java/tech/ula/BootReceiver.kt:
```diff
@@ -0,0 +1,21 @@
+import android.os.Build
+
+class BootReceiver : BroadcastReceiver() {
+    private val testAction = "tech.ula.intent.action.TEST_BOOT_COMPLETED"
```
This can actually be removed entirely.
I used this to test the broadcasts from adb shell and avoid restarting my device every time.

hugorosario commented on app/src/main/java/tech/ula/MainActivity.kt:
```diff
@@ -150,14 +151,13 @@ class MainActivity : AppCompatActivity(), SessionListFragment.SessionSelection,
                     destination.label == getString(R.string.sessions) ||
-                    destination.label == getString(R.string.apps) ||
-                    destination.label == getString(R.string.filesystems)
+                            destination.label == getString(R.string.apps) ||
```
Will do.

hugorosario commented on app/src/main/java/tech/ula/MainActivity.kt:
```diff
@@ -304,12 +302,24 @@ class MainActivity : AppCompatActivity(), SessionListFragment.SessionSelection,
-            is IllegalState -> { handleIllegalState(newState) }
-            is UserInputRequiredState -> { handleUserInputState(newState) }
-            is ProgressBarUpdateState -> { handleProgressBarUpdateState(newState) }
+            is CanOnlyStartSingleSession -> {
```
Will do. I think Android studio auto-indentation messed this up.

hugorosario commented on app/src/main/java/tech/ula/MainActivity.kt:
```diff
@@ -447,7 +456,7 @@ class MainActivity : AppCompatActivity(), SessionListFragment.SessionSelection,
             "playStoreMissingForClient" ->
                 displayGenericErrorDialog(this, R.string.alert_need_client_app_title,
-                    R.string.alert_need_client_app_message)
+                        R.string.alert_need_client_app_message)
```
Will do.

hugorosario commented on app/src/main/java/tech/ula/MainActivity.kt:
```diff
@@ -574,7 +581,7 @@ class MainActivity : AppCompatActivity(), SessionListFragment.SessionSelection,
             layout_progress.animation = inAnimation
 
-            layout_progress.visibility = View.VISIBLE
+            layout_progress.visibility = VISIBLE
```
Will do.

hugorosario commented on app/src/main/java/tech/ula/MainActivity.kt:
```diff
@@ -592,7 +599,7 @@ class MainActivity : AppCompatActivity(), SessionListFragment.SessionSelection,
         outAnimation.duration = 200
         layout_progress.animation = outAnimation
-        layout_progress.visibility = View.GONE
+        layout_progress.visibility = GONE
```
Will do

hugorosario commented on app/src/main/java/tech/ula/MainActivity.kt:
```diff
@@ -687,7 +691,7 @@ class MainActivity : AppCompatActivity(), SessionListFragment.SessionSelection,
 
                 val xsdlSupportedText = customDialog.findViewById<TextView>(R.id.text_xsdl_version_supported_description)
-                xsdlSupportedText.visibility = View.VISIBLE
+                xsdlSupportedText.visibility = VISIBLE
```
Will do

hugorosario commented on app/src/main/java/tech/ula/ServerService.kt:
```diff
@@ -27,6 +30,9 @@ class ServerService : Service() {
 
     private lateinit var broadcaster: LocalBroadcastManager
 
+    private var wakeLock: PowerManager.WakeLock? = null
```
1 - Good suggestion. Will look into this.
2 - I tried that way but I got a warning that it could cause a potential memory leak and should use application context instead.
2 - Will do but according to the documentation, the tag should be delimited by a colon.

"use a unique prefix delimited by a colon for your app/library (e.g. gmail:mytag) to make it easier to understand where the wake locks comes from. This namespace will also avoid collision for tags inside your app coming from different libraries which will make debugging easier."

So, maybe "Ula:ServerService"?

MatthewTighe replied:
2. If it's warning about memory leaks you can leave `applicationContext` as is. 
3. That seems like a good choice.

hugorosario commented on app/src/main/java/tech/ula/ServerService.kt:
```diff
@@ -80,17 +85,100 @@ class ServerService : Service() {
 
-        return Service.START_STICKY
+        val action = intent?.action
+        when (action) {
```
Will do

hugorosario commented on app/src/main/java/tech/ula/ServerService.kt:
```diff
@@ -80,17 +85,100 @@ class ServerService : Service() {
+                releaseWakeLocks()
+            }
+        }
+        return START_STICKY
```
Will do

hugorosario commented on app/src/main/java/tech/ula/ServerService.kt:
```diff
@@ -80,17 +85,100 @@ class ServerService : Service() {
+     * Fetch all sessions marked to start on boot and start them if they are not already active.
+     */
+    private fun autoStartSessions() {
+        CoroutineScope(Dispatchers.Default).launch {
```
I will look into this and change the implementation accordingly.

hugorosario replied:
@MatthewTighe I have been thinking about the different modes of setup that a user might need and came up with this set of options.
I will follow the same layout flow of the filesystem_edit fragment and make this options available under the "Show advanced options" toggle.
Some validation will be required for some cases (eg: There is no need to launch the client if the server is disabled.)

![image](https://user-images.githubusercontent.com/4238591/57003338-c1725f80-6bbd-11e9-8967-da0017fd8130.png)

MatthewTighe replied:
One more thing to think about here: There's two places that a command being executed on "start" could be executed if a server is also being started. 
1. _Before_ starting the server. I think the best way to run the command in this case would be to call it _from_ the script that starts the server. This would be a little work on our end, but it would mean that the name of the command (or better the location of a script containing any number of commands) would need to be passed as an environment variable to the `BusyboxExecutor`.
2. _After_ starting the server and having a client connect. This one is easier, and can just mirror the way that we run scripts when starting apps. That is to put them as the last entry in `/etc/profile.d` such that they execute automatically on client login.

hugorosario replied:
Will look into this after improving the current implementation.

hugorosario commented on app/src/main/java/tech/ula/ServerService.kt:
```diff
@@ -80,17 +85,100 @@ class ServerService : Service() {
+            var startedAnySession = false
+            for (session in sessions)
+                if (!session.active) {
+                    startSession(session, false)
```
The default value for the boolean value is True Wich means that for every other call to started soon, we will wait for the server and launch a client.

hugorosario replied:
This parameter will be removed on the new implementation I am working on.

hugorosario commented on app/src/main/java/tech/ula/ServerService.kt:
```diff
@@ -80,17 +85,100 @@ class ServerService : Service() {
+        val wm = applicationContext.getSystemService(Context.WIFI_SERVICE) as WifiManager
+        wifiLock = wm.createWifiLock(WifiManager.WIFI_MODE_FULL_HIGH_PERF, "ula:Service")
+        wifiLock?.acquire()
+        notificationManager.updateNotification(true)
```
Will do

hugorosario commented on app/src/main/java/tech/ula/ServerService.kt:
```diff
@@ -80,17 +85,100 @@ class ServerService : Service() {
+     * When the task is stopped, kill the server only if there are no startOnBoot sessions running.
+     */
+    override fun onTaskRemoved(rootIntent: Intent?) {
+        super.onTaskRemoved(rootIntent)
```
Good suggestion. It's a much better approach.

hugorosario commented on app/src/main/java/tech/ula/ServerService.kt:
```diff
@@ -110,23 +198,28 @@ class ServerService : Service() {
-    private suspend fun startSession(session: Session) {
-        startForeground(NotificationUtility.serviceNotificationId, notificationManager.buildPersistentServiceNotification())
+    private suspend fun startSession(session: Session, autoStartClient: Boolean = true) {
+        startForeground(NotificationUtility.serviceNotificationId, notificationManager.buildPersistentServiceNotification(wakeLock != null))
```
Will do and I will use wakeLock.isHeld() result instead of just checking nullability.

hugorosario commented on app/src/main/java/tech/ula/utils/NotificationUtility.kt:
```diff
@@ -38,19 +39,35 @@ class NotificationUtility(val context: Context) {
         val pendingSessionListIntent = PendingIntent
                 .getActivity(context, 0, sessionListIntent, 0)
 
+        var title = serviceNotificationTitle
```
I will change all the hardcoded strings into string resources. Not a problem.

hugorosario commented on app/src/main/java/tech/ula/utils/NotificationUtility.kt:
```diff
@@ -38,19 +39,35 @@ class NotificationUtility(val context: Context) {
+                .setContentTitle(title)
                 .setContentText(serviceNotificationDescription)
-                .setPriority(NotificationCompat.PRIORITY_MAX)
+                .setPriority(if (wakeLockHeld) NotificationCompat.PRIORITY_MAX else NotificationCompat.PRIORITY_LOW)
```
Yes. It can be always MAX priority. 
I don't think there is a problem.
It was done like this because the Termux notification has the same logic.

hugorosario commented on app/src/main/java/tech/ula/utils/NotificationUtility.kt:
```diff
@@ -38,19 +39,35 @@ class NotificationUtility(val context: Context) {
+
+        val newWakeAction = if (wakeLockHeld) "wakerelease" else "wakelock"
+        val toggleWakeLockIntent = Intent(context, ServerService::class.java).setAction(newWakeAction)
+        val actionTitle = if (wakeLockHeld) "Release Wakelock" else "Acquire Wakelock"
```
Will do

hugorosario commented:
Thank you for taking the time to review this.
I realize that this changes do not take into account the fact that UserLAnd only allows one active session.
Which means there is no need to have a startOnBoot property on the sessions and no need to iterate and launch all the sessions with this property.

As suggested, I will remove the property from the sessions and save the current "startOnBoot" session in a shared preference.
If a user attempts to set another session as a "startOnBoot" I will display a confirmation explaining that only one session can be started on boot.
I will also create sharedprefs for defining the autostart script/command to be executed (I was already planning to do this).

hugorosario commented:
@MatthewTighe I have fixed most of the issues and refactored the implementation of startOnBoot.
It might not be perfect yet, but it should be better.

MatthewTighe commented on app/src/main/java/tech/ula/ui/SessionEditFragment.kt:
```diff
@@ -189,6 +234,22 @@ class SessionEditFragment : Fragment() {
 
+        // If another session has the startOnBoot property, warn the user
+        if (checkbox_start_boot.isChecked) {
+            if (sessionPreferences.getStartOnBootSession() != session.name && sessionPreferences.getStartOnBootSession().isNotEmpty()) {
```
Here's another candidate for using `session.id` in the conditional instead of `session.name`.

MatthewTighe commented on app/src/main/java/tech/ula/ui/SessionEditFragment.kt:
```diff
@@ -189,6 +234,22 @@ class SessionEditFragment : Fragment() {
+                    insertSession()
+                }
+                return
+            } else
```
Same style issue with the `else` statement here.

MatthewTighe commented on app/src/main/java/tech/ula/utils/AndroidUtility.kt:
```diff
@@ -59,6 +59,21 @@ fun displayGenericErrorDialog(activity: Activity, titleId: Int, messageId: Int,
+            .setMessage(messageId)
+            .setTitle(titleId)
+            .setPositiveButton(R.string.button_positive) {
+                dialog, _ ->
```
I'd like for the style of the parameters for these lambdas to match, at least.

MatthewTighe commented on app/src/main/java/tech/ula/model/entities/Session.kt:
```diff
@@ -32,11 +32,14 @@ data class Session(
         return "Session(id=$id, name=$name, filesystemId=$filesystemId, filesystemName=" +
                 "$filesystemName, active=$active, serviceType=$serviceType, port=$port, pid=" +
-                "$pid, isAppsSession=$isAppsSession), startOnBoot=$startOnBoot"
+                "$pid, isAppsSession=$isAppsSession), startCommand=$startCommand, " +
```
Would you move `startCommand` to be the last entry in the string? I'm worried if it's long it could obscure valuable information in our logs.

MatthewTighe commented on app/src/main/java/tech/ula/utils/ServerUtility.kt:
```diff
@@ -57,13 +55,19 @@ class ServerUtility(
+        val profileScript = File(applicationFilesDirPath, "$filesystemDirName/support/userland_profile.sh")
+        profileScript.copyTo(startScript, overwrite = true)
+        startScript.appendText("\n${session.startCommand}")
+        startScript.setExecutable(true, false)
```
I'd like to see named parameters for the boolean values in `setExecutable` like we talked about previously. Just helps people reading the code understand the call without needing to look up documentation.

hugorosario replied:
Not possible in this case. 
Only possible when calling Kotlin functions and the File class is Java.

MatthewTighe replied:
It could be useful to define the booleans as variables before the call then. 

```
val isExecutable = true
val isOwnerOnly = false
startScript.setExecutable(isExecutable, isOwnerOnly)
```

That's pretty nitpicky though, so not a required change.

MatthewTighe commented on app/src/main/java/tech/ula/utils/ServerUtility.kt:
```diff
@@ -57,13 +55,19 @@ class ServerUtility(
+            return
         val filesystemDirName = session.filesystemId.toString()
-        // TODO create new session parameter and turn this into a user-defined command
         val command = "/support/autostart.sh"
```
I think at this point, `command` is a slightly disingenuous name for this variable. I think it might make the most sense to just remove it and change `startScript` to include that part of the path.

MatthewTighe commented on app/src/main/java/tech/ula/ui/SessionEditFragment.kt:
```diff
@@ -189,6 +234,22 @@ class SessionEditFragment : Fragment() {
+        }
+
+        // If this was the startOnBoot session and the user unchecked the property, clear the flag
+        if (!checkbox_start_boot.isChecked && sessionPreferences.getStartOnBootSession() == session.name)
```
Another candidate for checking `id.`

MatthewTighe commented on app/src/main/java/tech/ula/utils/ServerUtility.kt:
```diff
@@ -57,13 +55,19 @@ class ServerUtility(
+     * Creates a new script based on userland_profile.sh and adds the user-defined command to the script
      */
-    private fun runBootScript(session: Session) {
+    fun executeStartCommand(session: Session) {
```
I'd like to see at least one test for this function, mostly for regression purposes.

hugorosario replied:
You are going to have to help me on that because I never wrote a single unit test in my life.
What should we test for in this function?

MatthewTighe replied:
Sure, I can help you with that. Generally when writing unit tests, the minimum criteria is to write a test for each possible branch. So one test would just assert that if `executeStartCommand` is called with a `session` that had an empty `startCommand`, that it would return early. There are actually 2 ways to assert that for this case:

1. Assert that the `startScript` file is never overwritten.
2. Verify that `busyboxExecutor` is never called.

If you look in `ServerUtilityTest.kt` you can start seeing a bit about how these are tested. This should also start making it more clear why the dependencies that are injected into `ServerUtility` are injected into it: because it makes the class unit-testable. By provided mocks for the objects it uses, and a temporary folder to take the place of `applicationFilesDir`, we can guarantee that the tests will run as a unit anywhere we can run the JVM nd not just on the Android platform itself.

So onto writing the test itself:

```
    @Test
    fun `Calling executeStartCommand on a session without a start command returns early`() {
        // Given
        val filesystemId = 1L
        val session = Session(id = 0, filesystemId = filesystemId)
        val startScriptFile = File("${tempFolder.root}/$filesystemId/support/autostart.sh")
        val profileScriptFile = File("${tempFolder.root}/$filesystemId/support/userland_profile.sh")

        val startText = "start"
        startScriptFile.writeText(startText)

        val profileText = "profile"
        profileScriptFile.writeText(profileText)

        // When
        serverUtility.executeStartCommand(session)

        // Then
        val expectedCommand = "/support/autostart.sh"
        val expectedCommandShouldTerminate = false
        assertEquals(startText, startScriptFile.readText())
        verify(mockBusyboxExecutor, never()).executeProotCommand(
            eq(expectedCommand),
            eq(filesystemId.toString()),
            eq(expectedCommandShouldTerminate),
            anyOrNull(),
            anyOrNull()
        )
    }
```

Unit tests are typically written in a 'given, when, then` style. I've left some comments to show those stages for this test.

This test would pass, but in writing it it helped my understanding of some ways that the implementation of this function could be improved.

First of all, the function as it is written has an underlying (and hidden) assumption about a precondition. Namely, that `profileScript` already exists. That's a pretty safe assumption given the structure of the app as a whole, but it's worth accounting for it as a failure mode anyway. If the file doesn't exist, this function will throw an exception which isn't caught and handled explicitly elsewhere. That's probably a fine mode of failure in this case, but it leads nicely into another point: what happens if the command fails to execute? Right now, that would just result in a silent failure, which would likely lead to confusion on the part of the user. If the session is also going to start a server, that server and client would still be started. It would probably be better to stop starting the session and alert the user of the failure through a dialog. Since we would then have the infrastructure set up to alert of failures, maybe it's worth informing them of the failure case of the source `profileScript` not existing.

You can then write tests for the successful and failure cases pretty easily.

In terms of code structure, I think this function might actually be better placed in `BusyboxExecutor` itself. Nothing that this function does is explicitly related to servers. That being said, writing tests for it there would actually probably be harder. Feel free to leave it here for now, but maybe add a `TODO` comment to refactor it later.

If you're so inclined, [here](https://www.philosophicalhacker.com/2015/04/17/why-android-unit-testing-is-so-hard-pt-1/) is a series of blog posts that helped me deepen my understanding of the difficulties of unit testing on the Android platform.

If you have any more questions leave them here or find me in slack!

hugorosario replied:
Thank you for taking the time to explain this.
This is valuable information for me and I will take a look at the blog you linked.
The executeStartCommand function can definitely be improved.

We can add a few checks and I can think of a few ways to alert the user about the failure of execution :

1 - Send a broadcast from the service and have the MainActivity handle it and showing an error dialog. This would take care of sessions that are being started manually by the user, but would be useless on boot sessions because there is no activity listening.

2 - A new notification (or modifying the persistent one) to alert the user.

3 - We could send an intent to the MainActivity from the Service.
Something like :
```
Intent dialogIntent = new Intent(this, MyActivity.class);  
dialogIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);  
dialogIntent.putExtra("startBootFailed", true);
startActivity(dialogIntent);
```
When launching from a service, we need to use the FLAG_ACTIVITY_NEW_TASK.
This method works for both cases (user-initiated sessions and boot sessions) but needs to be handled differently in case the activity is already running. In this case, it will just be brought to the front without recreating the activity and will fire the onNewIntent() callback.

MatthewTighe replied:
I'm fine with either solution. Using a notification should probably create a new one rather than reuse the persistent so it's clear something happened. I imagine the notification would just lead to the app being opened anyway though, so maybe we just skip the middleman. Up to you.

hugorosario replied:
I will probably use option 3.
I have been busy with work but I plan on finishing the remaining issues this week.

MatthewTighe commented on app/src/main/java/tech/ula/ui/SessionEditFragment.kt:
```diff
@@ -12,18 +12,27 @@ import android.view.* // ktlint-disable no-wildcard-imports
     private lateinit var activityContext: Activity
 
+    private val sessionPreferences by lazy {
+        SessionPreferences(activityContext.defaultSharedPreferences)
```
e.g. `activityContext.getSharedPreferences("startOnBootSession", Context.MODE_PRIVATE)`

MatthewTighe commented:
@hugorosario Any updates here? Just want to know if it's still active or not.

hugorosario commented:
> @hugorosario Any updates here? Just want to know if it's still active or not.

Hey there. Sorry about the delay. I have been on vacations last week.
Will be getting back to this tomorrow.

MatthewTighe commented:
No worries, glad to hear you're still interested.

brianddk commented:
Bit of a tangent, but does anyone know the [android `am start` commands](https://stackoverflow.com/a/4567928/4634229) that need to be fired to start a UserLAnd session?  I have my Android instance hooked to adb so if I knew the right ActivityName, ActionName or Intent I could script it through adb if my uptime-monitor found it down.

MatthewTighe commented:
@brianddk That's pretty off-topic for this PR. Consider opening an issue.

rhlobo commented:
Great work @hugorosario ! This will be huge for me. @MatthewTighe , when are you planning to merge and publish this? Something I can help?

hugorosario commented:
> Great work @hugorosario ! This will be huge for me. @MatthewTighe , when are you planning to merge and publish this? Something I can help?

I am so sorry for not completing the requested changes but I have been very busy with work and it has been difficult to find a moment to get back to this.
If anyone wants to pick this up and finish the task, go for it.

MatthewTighe commented:
I can add this to our backlog, but it won't become an immediate priority. Feel free to continue working on it if you find time, and I'll let you know if/when I pick it up.

rhlobo commented:
Don´t worry @hugorosario <https://github.com/hugorosario>, take your time.

I will check if this is something I may be able to do: Am kind of rusty, 5
years away from programming (and never did anything for android).

@MatthewTighe <https://github.com/MatthewTighe>, can you throw at me some
of the keywords of what I have to install to get going with the dev env?
Only the Android SDK and my editor?

On Tue, Jun 18, 2019 at 3:15 PM Matthew Tighe <notifications@github.com>
wrote:

> I can add this to our backlog, but it won't become an immediate priority.
> Feel free to continue working on it if you find time, and I'll let you know
> if/when I pick it up.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/CypherpunkArmory/UserLAnd/pull/801?email_source=notifications&email_token=AAJKL4RLMQ7LZZD3RLZVH2TP3EQ43A5CNFSM4HJKRXV2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODX7QELA#issuecomment-503251500>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAJKL4TIP6TX62ZL2LQIY2DP3EQ43ANCNFSM4HJKRXVQ>
> .
>

MatthewTighe commented:
@rhlobo If you're considering doing some Android development, I would strongly recommend just doing it through Android Studio. It'll guide you through the setup process and have useful error messages for missing SDK components. Otherwise yeah, you'll need to configure it manually. Google will be a better resource there than I am.

rhlobo commented:
Ok, thats what I needed to know

On Tue, Jun 18, 2019 at 7:04 PM Matthew Tighe <notifications@github.com>
wrote:

> @rhlobo <https://github.com/rhlobo> If you're considering doing some
> Android development, I would strongly recommend just doing it through
> Android Studio. It'll guide you through the setup process and have useful
> error messages for missing SDK components. Otherwise yeah, you'll need to
> configure it manually. Google will be a better resource there than I am.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/CypherpunkArmory/UserLAnd/pull/801?email_source=notifications&email_token=AAJKL4T2EAEB4QEOXXMU2HDP3FLW5A5CNFSM4HJKRXV2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYADFGA#issuecomment-503329432>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAJKL4VISYHULAHHEWPRPBTP3FLW5ANCNFSM4HJKRXVQ>
> .
>

rhlobo commented:
I've got the code, but as never done anything on Android this is a bigger task than initially I've imagined. I wont be able to help much at this time. Sorry

rhlobo commented:
If my contribution is accepted, I think next step is to *send an intent to the MainActivity from the Service* and UI.

rhlobo commented:
Done some more work. @MatthewTighe , whenever you have some time, I would appreciate if you can check the [PR](https://github.com/hugorosario/UserLAnd/pull/1) I am doing to @hugorosario original PR, and let me know if I going in the right direction.

Cheers

MatthewTighe commented:
@rhlobo It certainly looks like you're headed in the right direction! That being said, I have a couple concerns:
- I think UserLAnd has diverged far enough from where this PR was based that it might be worth forking master from this repo. You could selectively reapply changes, or rewrite the PR using the original concepts. By starting from a clean fork, you can take advantage of improvements to architecture and file naming that I think are relevant to this PR.
- Rather than making PRs to @hugorosario, it would be helpful if you made PRs directly to this repo as well. This helps keep the conversation contained, as well as moving responsibility from @hugorosario to us at UserLAnd to request the needed changes and ultimately accept the PR.

That being said, I'm operating under the assumption that @hugorosario has decided not to continue working on this contribution. If that is not the case, the two of you can obviously collaborate in whatever way works best.

MatthewTighe commented:
I am going to close this PR since it hasn't seen an update in a while. Feel free to keep the conversation going here, and ping me on our [slack](https://communityinviter.com/apps/userlandtech/userland).

rhlobo commented:
Hey, first and foremost, thank you for your answer. I did not reply because
had nothing usefull to add.

Anyway, since last weekend I had no time to restart this PR again (I left
programming to take care of family business, CEO of 300+ employees). I
realized this should be completed fast, or else commits on master will make
the effort useless, so I will try to do it at once, maybe this weekend.
When this happens, I will take your advice and make a direct PR.

Cheers

On Fri, Jul 19, 2019 at 4:45 PM Matthew Tighe <notifications@github.com>
wrote:

> I am going to close this PR since it hasn't seen an update in a while.
> Feel free to keep the conversation going here, and ping me on our slack
> <https://communityinviter.com/apps/userlandtech/userland>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/CypherpunkArmory/UserLAnd/pull/801?email_source=notifications&email_token=AAJKL4VGS4K7VMED5BZOWNLQAIKWTA5CNFSM4HJKRXV2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD2MSWUA#issuecomment-513354576>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAJKL4QZKMI2OJ6EJNHWGJLQAIKWTANCNFSM4HJKRXVQ>
> .
>
